File: java/src/main/java/vsdk/toolkit/skin/RayHit.java
package vsdk.toolkit.skin;

import vsdk.toolkit.common.Error;
import vsdk.toolkit.common.linealAlgebra.CoordinateSystem;
import vsdk.toolkit.common.linealAlgebra.Vector2Dd;
import vsdk.toolkit.common.linealAlgebra.Vector3D;
import vsdk.toolkit.material.Material;
import vsdk.toolkit.material.PhongBidirectionalScatteringDistributionFunction;
import vsdk.toolkit.material.PhongEmittanceDistributionFunction;
import vsdk.toolkit.material.RayHitFlag;

public class RayHit {
    private Vector3D point;
    private Patch patch;
    private Vector3D texCoord;
    private Vector3D geometricNormal;
    private Material material;
    private CoordinateSystem shadingFrame;
    private Vector2Dd uv;
    private int flags;

    public RayHit() {
        point = new Vector3D();
        patch = null;
        texCoord = new Vector3D();
        geometricNormal = new Vector3D();
        material = null;
        shadingFrame = new CoordinateSystem();
        uv = new Vector2Dd();
        flags = 0;
    }

    private boolean computeUv(Vector2Dd inUv) {
        if ((flags & RayHitFlag.UV) != 0) {
            inUv.u = uv.u;
            inUv.v = uv.v;
            return true;
        }

        if (((flags & RayHitFlag.PATCH) != 0) && ((flags & RayHitFlag.POINT) != 0) && patch != null) {
            double[] u = new double[] {0.0};
            double[] v = new double[] {0.0};
            patch.uv(point, u, v);
            uv.u = u[0];
            uv.v = v[0];
            inUv.u = uv.u;
            inUv.v = uv.v;
            flags |= RayHitFlag.UV;
            return true;
        }

        return false;
    }

    private boolean hitInitialised() {
        return (((flags & RayHitFlag.PATCH) != 0) || ((flags & RayHitFlag.GEOMETRY) != 0))
            && ((flags & RayHitFlag.POINT) != 0)
            && ((flags & RayHitFlag.GEOMETRIC_NORMAL) != 0)
            && ((flags & RayHitFlag.MATERIAL) != 0)
            && ((flags & RayHitFlag.DISTANCE) != 0);
    }

    private boolean pointShadingFrame(Vector3D inX, Vector3D inY, Vector3D inZ) {
        boolean success = false;

        if (!hitInitialised()) {
            Error.warning("pointShadingFrame", "uninitialised hit structure");
            return false;
        }

        if (material != null && material.getBsdf() != null) {
            success = PhongBidirectionalScatteringDistributionFunction.bsdfShadingFrame(this, inX, inY, inZ);
        }

        if (!success && material != null && material.getEdf() != null) {
            success = PhongEmittanceDistributionFunction.edfShadingFrame(this, inX, inY, inZ);
        }

        if (!success && computeUv(uv) && patch != null) {
            Vector3D x = inX != null ? inX : new Vector3D();
            Vector3D y = inY != null ? inY : new Vector3D();
            Vector3D z = inZ != null ? inZ : new Vector3D();
            patch.interpolatedFrameAtUv(uv.u, uv.v, x, y, z);

            if (inZ != null) {
                inZ.copy(z);
            }
            if (inX != null) {
                inX.copy(x);
            }
            if (inY != null) {
                inY.copy(y);
            }
            success = true;
        }

        return success;
    }

    public boolean init(Patch inPatch, Vector3D inPoint, Vector3D inGeometryNormal, Material inMaterial) {
        flags = 0;
        patch = inPatch;
        if (inPatch != null) {
            flags |= RayHitFlag.PATCH;
        }

        if (inPoint != null) {
            point.copy(inPoint);
            flags |= RayHitFlag.POINT;
        }

        if (inGeometryNormal != null) {
            geometricNormal.copy(inGeometryNormal);
            flags |= RayHitFlag.GEOMETRIC_NORMAL;
        }

        material = inMaterial;
        flags |= RayHitFlag.MATERIAL;
        flags |= RayHitFlag.DISTANCE;

        Vector3D localNormal = new Vector3D();
        localNormal.set(0.0f, 0.0f, 0.0f);
        texCoord.copy(localNormal);
        shadingFrame.setX(localNormal);
        shadingFrame.setY(localNormal);
        shadingFrame.setZ(localNormal);
        uv.u = 0.0;
        uv.v = 0.0;

        return hitInitialised();
    }

    public boolean getTexCoord(Vector3D outTexCoord) {
        if ((flags & RayHitFlag.TEXTURE_COORDINATE) != 0) {
            outTexCoord.copy(texCoord);
            return true;
        }

        if (!computeUv(uv)) {
            return false;
        }

        if ((flags & RayHitFlag.PATCH) != 0 && patch != null) {
            texCoord = patch.textureCoordAtUv(uv.u, uv.v);
            outTexCoord.copy(texCoord);
            flags |= RayHitFlag.TEXTURE_COORDINATE;
            return true;
        }

        return false;
    }

    public boolean shadingNormal(Vector3D inNormal) {
        if (((flags & RayHitFlag.SHADING_FRAME) != 0) || ((flags & RayHitFlag.NORMAL) != 0)) {
            inNormal.copy(shadingFrame.getZ());
            return true;
        }

        Vector3D localNormal = shadingFrame.getZ();
        if (!pointShadingFrame(null, null, localNormal)) {
            return false;
        }

        flags |= RayHitFlag.NORMAL;
        shadingFrame.setZ(localNormal);
        inNormal.copy(shadingFrame.getZ());
        return true;
    }

    public Patch getPatch() {
        return patch;
    }

    public void setPatch(Patch inPatch) {
        patch = inPatch;
    }

    public Vector3D getPoint() {
        return point;
    }

    public void setPoint(Vector3D position) {
        point = new Vector3D(position.x, position.y, position.z);
    }

    public void setGeometricNormal(Vector3D inNormal) {
        geometricNormal = new Vector3D(inNormal.x, inNormal.y, inNormal.z);
    }

    public void setMaterial(Material inMaterial) {
        material = inMaterial;
    }

    public Vector2Dd getUv() {
        return uv;
    }

    public void setUv(Vector2Dd inUv) {
        uv.u = inUv.u;
        uv.v = inUv.v;
    }

    public void setUv(double inU, double inV) {
        uv.u = inU;
        uv.v = inV;
    }

    public int getFlags() {
        return flags;
    }

    public void setFlags(int inFlags) {
        flags = inFlags;
    }

    public boolean setShadingFrame(CoordinateSystem frame) {
        if ((flags & RayHitFlag.SHADING_FRAME) != 0) {
            frame.setX(shadingFrame.getX());
            frame.setY(shadingFrame.getY());
            frame.setZ(shadingFrame.getZ());
            return true;
        }

        Vector3D shadingX = shadingFrame.getX();
        Vector3D shadingY = shadingFrame.getY();
        Vector3D shadingZ = shadingFrame.getZ();

        if (!pointShadingFrame(shadingX, shadingY, shadingZ)) {
            return false;
        }

        shadingFrame.setX(shadingX);
        shadingFrame.setY(shadingY);
        shadingFrame.setZ(shadingZ);
        flags |= RayHitFlag.SHADING_FRAME | RayHitFlag.NORMAL;

        frame.setX(shadingFrame.getX());
        frame.setY(shadingFrame.getY());
        frame.setZ(shadingFrame.getZ());
        return true;
    }

    public Vector3D getNormal() {
        return shadingFrame.getZ();
    }

    public void setNormal(Vector3D n) {
        shadingFrame.setZ(n);
    }

    public CoordinateSystem getShadingFrame() {
        return shadingFrame;
    }

    public Material getMaterial() {
        return material;
    }

    public Vector3D getGeometricNormal() {
        return geometricNormal;
    }

    public void setShadingFrame(Vector3D inX, Vector3D inY, Vector3D inZ) {
        shadingFrame.setX(inX);
        shadingFrame.setY(inY);
        shadingFrame.setZ(inZ);
    }
}

File: typescript/src/main/vsdk/toolkit/skin/RayHit.ts
import { Error as VsdkError } from "../common/Error";
import { CoordinateSystem } from "../common/linealAlgebra/CoordinateSystem";
import { Vector2Dd } from "../common/linealAlgebra/Vector2Dd";
import { Vector3D } from "../common/linealAlgebra/Vector3D";
import { Material } from "../material/Material";
import { PhongBidirectionalScatteringDistributionFunction } from "../material/PhongBidirectionalScatteringDistributionFunction";
import { PhongEmittanceDistributionFunction } from "../material/PhongEmittanceDistributionFunction";
import { RayHitFlag } from "../skin/RayHitFlag";
import type { Patch } from "./Patch";

export class RayHit {
  private point: Vector3D;
  private patch: Patch | null;
  private texCoord: Vector3D;
  private geometricNormal: Vector3D;
  private material: Material | null;
  private shadingFrame: CoordinateSystem;
  private uv: Vector2Dd;
  private flags: number;

  public constructor() {
    this.point = new Vector3D();
    this.patch = null;
    this.texCoord = new Vector3D();
    this.geometricNormal = new Vector3D();
    this.material = null;
    this.shadingFrame = new CoordinateSystem();
    this.uv = new Vector2Dd();
    this.flags = 0;
  }

  private hitInitialised(): boolean {
    return (((this.flags & RayHitFlag.PATCH) !== 0) || ((this.flags & RayHitFlag.GEOMETRY) !== 0))
      && ((this.flags & RayHitFlag.POINT) !== 0)
      && ((this.flags & RayHitFlag.GEOMETRIC_NORMAL) !== 0)
      && ((this.flags & RayHitFlag.MATERIAL) !== 0)
      && ((this.flags & RayHitFlag.DISTANCE) !== 0);
  }

  private computeUv(inUv: Vector2Dd): boolean {
    if ((this.flags & RayHitFlag.UV) !== 0) {
      inUv.u = this.uv.u;
      inUv.v = this.uv.v;
      return true;
    }

    if (((this.flags & RayHitFlag.PATCH) !== 0) && ((this.flags & RayHitFlag.POINT) !== 0) && this.patch !== null) {
      const u = [0.0];
      const v = [0.0];
      this.patch.uv(this.point, u, v);
      this.uv.u = u[0];
      this.uv.v = v[0];
      inUv.u = this.uv.u;
      inUv.v = this.uv.v;
      this.flags |= RayHitFlag.UV;
      return true;
    }

    return false;
  }

  private pointShadingFrame(inX: Vector3D | null, inY: Vector3D | null, inZ: Vector3D | null): boolean {
    let success = false;

    if (!this.hitInitialised()) {
      VsdkError.warning("pointShadingFrame", "uninitialised hit structure");
      return false;
    }

    if (this.material !== null && this.material.getBsdf() !== null) {
      success = PhongBidirectionalScatteringDistributionFunction.bsdfShadingFrame(this, inX, inY, inZ);
    }

    if (!success && this.material !== null && this.material.getEdf() !== null) {
      success = PhongEmittanceDistributionFunction.edfShadingFrame(this, inX, inY, inZ);
    }

    if (!success && this.computeUv(this.uv) && this.patch !== null) {
      const x = inX !== null ? inX : new Vector3D();
      const y = inY !== null ? inY : new Vector3D();
      const z = inZ !== null ? inZ : new Vector3D();
      this.patch.interpolatedFrameAtUv(this.uv.u, this.uv.v, x, y, z);

      if (inX !== null) {
        inX.copy(x);
      }
      if (inY !== null) {
        inY.copy(y);
      }
      if (inZ !== null) {
        inZ.copy(z);
      }
      success = true;
    }

    return success;
  }

  public init(
    inPatch: Patch | null,
    inPoint: Vector3D | null,
    inGeometryNormal: Vector3D | null,
    inMaterial: Material | null
  ): boolean {
    this.flags = 0;
    this.patch = inPatch;
    if (inPatch !== null) {
      this.flags |= RayHitFlag.PATCH;
    }

    if (inPoint !== null) {
      this.point.copy(inPoint);
      this.flags |= RayHitFlag.POINT;
    }

    if (inGeometryNormal !== null) {
      this.geometricNormal.copy(inGeometryNormal);
      this.flags |= RayHitFlag.GEOMETRIC_NORMAL;
    }

    this.material = inMaterial;
    this.flags |= RayHitFlag.MATERIAL;
    this.flags |= RayHitFlag.DISTANCE;

    const localNormal = new Vector3D();
    localNormal.set(0.0, 0.0, 0.0);
    this.texCoord.copy(localNormal);
    this.shadingFrame.setX(localNormal);
    this.shadingFrame.setY(localNormal);
    this.shadingFrame.setZ(localNormal);
    this.uv.u = 0.0;
    this.uv.v = 0.0;

    return this.hitInitialised();
  }

  public getTexCoord(outTexCoord: Vector3D): boolean {
    if ((this.flags & RayHitFlag.TEXTURE_COORDINATE) !== 0) {
      outTexCoord.copy(this.texCoord);
      return true;
    }

    if (!this.computeUv(this.uv)) {
      return false;
    }

    if ((this.flags & RayHitFlag.PATCH) !== 0 && this.patch !== null) {
      this.texCoord = this.patch.textureCoordAtUv(this.uv.u, this.uv.v);
      outTexCoord.copy(this.texCoord);
      this.flags |= RayHitFlag.TEXTURE_COORDINATE;
      return true;
    }

    return false;
  }

  public shadingNormal(inNormal: Vector3D): boolean {
    if (((this.flags & RayHitFlag.SHADING_FRAME) !== 0) || ((this.flags & RayHitFlag.NORMAL) !== 0)) {
      inNormal.copy(this.shadingFrame.getZ());
      return true;
    }

    const localNormal = this.shadingFrame.getZ();
    if (!this.pointShadingFrame(null, null, localNormal)) {
      return false;
    }

    this.flags |= RayHitFlag.NORMAL;
    this.shadingFrame.setZ(localNormal);
    inNormal.copy(this.shadingFrame.getZ());
    return true;
  }

  public setShadingFrame(frame: CoordinateSystem): boolean;
  public setShadingFrame(inX: Vector3D, inY: Vector3D, inZ: Vector3D): void;
  public setShadingFrame(
    frameOrX: CoordinateSystem | Vector3D,
    inY?: Vector3D,
    inZ?: Vector3D
  ): boolean | void {
    if (frameOrX instanceof CoordinateSystem) {
      const frame = frameOrX;
      if ((this.flags & RayHitFlag.SHADING_FRAME) !== 0) {
        frame.setX(this.shadingFrame.getX());
        frame.setY(this.shadingFrame.getY());
        frame.setZ(this.shadingFrame.getZ());
        return true;
      }

      const shadingX = this.shadingFrame.getX();
      const shadingY = this.shadingFrame.getY();
      const shadingZ = this.shadingFrame.getZ();

      if (!this.pointShadingFrame(shadingX, shadingY, shadingZ)) {
        return false;
      }

      this.shadingFrame.setX(shadingX);
      this.shadingFrame.setY(shadingY);
      this.shadingFrame.setZ(shadingZ);
      this.flags |= RayHitFlag.SHADING_FRAME | RayHitFlag.NORMAL;

      frame.setX(this.shadingFrame.getX());
      frame.setY(this.shadingFrame.getY());
      frame.setZ(this.shadingFrame.getZ());
      return true;
    }

    this.shadingFrame.setX(frameOrX);
    this.shadingFrame.setY(inY as Vector3D);
    this.shadingFrame.setZ(inZ as Vector3D);
    return;
  }

  public getPatch(): Patch | null {
    return this.patch;
  }

  public setPatch(inPatch: Patch | null): void {
    this.patch = inPatch;
  }

  public getPoint(): Vector3D {
    return this.point;
  }

  public setPoint(position: Vector3D): void {
    this.point.copy(position);
  }

  public setGeometricNormal(inNormal: Vector3D): void {
    this.geometricNormal.copy(inNormal);
  }

  public setMaterial(inMaterial: Material | null): void {
    this.material = inMaterial;
  }

  public getUv(): Vector2Dd {
    return this.uv;
  }

  public setUv(inUv: Vector2Dd): void;
  public setUv(inU: number, inV: number): void;
  public setUv(inUvOrU: Vector2Dd | number, inV?: number): void {
    if (inUvOrU instanceof Vector2Dd) {
      this.uv.u = inUvOrU.u;
      this.uv.v = inUvOrU.v;
      return;
    }

    this.uv.u = inUvOrU;
    this.uv.v = inV as number;
  }

  public getFlags(): number {
    return this.flags;
  }

  public setFlags(inFlags: number): void {
    this.flags = inFlags;
  }

  public getNormal(): Vector3D {
    return this.shadingFrame.getZ();
  }

  public setNormal(n: Vector3D): void {
    this.shadingFrame.setZ(n);
  }

  public getShadingFrame(): CoordinateSystem {
    return this.shadingFrame;
  }

  public getMaterial(): Material | null {
    return this.material;
  }

  public getGeometricNormal(): Vector3D {
    return this.geometricNormal;
  }

  public copyFrom(other: RayHit | null): void {
    if (other === null) {
      return;
    }

    this.setPoint(other.getPoint());
    this.setPatch(other.getPatch());
    this.setGeometricNormal(other.getGeometricNormal());
    this.setMaterial(other.getMaterial());
    this.setUv(other.getUv());
    this.setFlags(other.getFlags());

    const frame = other.getShadingFrame();
    this.setShadingFrame(frame.getX(), frame.getY(), frame.getZ());
    const localTex = new Vector3D();
    if (other.getTexCoord(localTex)) {
      this.texCoord = localTex;
    }
  }
}
